File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    int r;
    int c;
    // i is row index, j is column index
    public void recursion(int i, int j, char[][] grid){
        if (i-1 > 0 && grid[i-1][j] == '1'){
            grid[i-1][j] = '0';
            recursion(i-1, j, grid);
        } 
        if (i+1 < r && grid[i+1][j] == '1'){
            grid[i+1][j] = '0';
            recursion(i+1, j, grid);
        }
        if (j-1 > 0 && grid[i][j-1] == '1'){
            grid[i][j-1] = '0';
            recursion(i, j-1, grid);
        }
        if (j+1 < c && grid[i][j+1] == '1'){
            grid[i][j+1] = '0';
            recursion(i, j+1, grid);
        }
        return;
        
    }
    public int numIslands(char[][] grid) {
        r = grid.length;
        c = grid[0].length;
        int count = 0;
        for (int i = 0; i < r; i++){
            for (int j = 0; j < c; j++){
                if (grid[i][j] == '1'){
                    count++;
                    grid[i][j] = '0';
                    recursion(i, j, grid);
                }
            }
        }
        return count;
    }
}

File: 0200-number-of-islands/0200-number-of-islands.cs
public class Solution {
    int[][] directions = new int[][]{
        new int[] {1,0},
        new int[] {0,1},
        new int[] {-1, 0},
        new int[] {0, -1}
    };
    public void Bfs(int i, int j, char[][] grid, int r_len, int c_len){
        Queue<int[]> q = new Queue<int[]>();
        grid[i][j] = '0';
        q.Enqueue(new int[]{i,j});
        
        while (q.Count() > 0){
            int[] point = q.Dequeue();
            foreach(int[] direction in directions){
                int r = direction[0] + point[0];
                int c = direction[1] + point[1];

                if (r >= 0 && c >= 0 && r < r_len && c < c_len && grid[r][c] == '1'){
                    q.Enqueue(new int[] {r, c});
                    grid[r][c] = '0';
                }
            }

        }
    } 

    public int NumIslands(char[][] grid) {

        int res = 0;
        int r_len = grid.Length;
        int c_len = grid[0].Length;

        for (int i = 0; i < r_len; i++){
            for (int j = 0; j < c_len; j++){
                if (grid[i][j] == '1'){
                    res++;
                    Bfs(i, j, grid, r_len, c_len);
                }
            }
        }
        return res;
    }
}